#Bài toán:Phát hiện lỗi sai và chấm điểm phát âm tiếng Anh bằng AI <br>



##Phương pháp và các bước xử lý bài toán
**Ý tưởng:** Tạo một bộ dữ liệu trainin gắn nhãn theo phiên âm(Phonemes), Từ đó sử dụng các mô hình học máy để nhận diện từng phiên âm trong file audio.<br>
**Các bước xử lý**:<br>
B1: Chuẩn bị dữ liệu làm và gắn nhãn. Từ bộ dữ liệu TIMIT(EU,US,UK), CMU(UK), L2-corpus(GLOBAL), gắn nhãn và lọc dữ liệu không phải tiếng Anh...<br>
B2: Data precessing,data cleaning<br>
B3: Mã hoá dữ liệu theo phương pháp 2-gram,3-gram,Wav2Vec<br>
B4: Mapping phiên âm<br>
B5: Hậu xử lý mô hình<br>

# Khởi Tạo Môi Trường

In [ ]:
!pip install transformers
!pip install jiwer
!pip install phonemizer
!pip search espeak
!pip install py-espeak-ng
!sudo apt-get install python-espeak
!sudo apt-get update && sudo apt-get install espeak
!pip install espeak

In [ ]:
from transformers import Wav2Vec2Processor, Wav2Vec2ForCTC
from transformers import Wav2Vec2PhonemeCTCTokenizer
import torch
import re
import librosa
import os
import pandas as pd
from jiwer import wer,cer
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

# Load file audio#

In [ ]:
class ReadLabFile():
    def __init__(self,path):
        self.path = path
        self.path_labs,self.path_wavs = self.path_wavs_1()
        
    def path_wavs_1(self):
        paths = os.listdir(self.path)
        path_wavs=[]
        path_labs=[]
        for path in paths:
            if '.wav' in path:
                path_wavs.append(path)
            if '.lab' in path:
                path_labs.append(path)
        path_labs.sort()
        path_wavs.sort()
        return path_labs,path_wavs
    
    def read_speech_file(self,value_drop=['pau']):
        paths = self.path_labs
        val = []
        valu_=[]
        for path in paths:
            values = list(pd.read_csv('abc'+'/'+path, sep = ' ')['#'])
            val.append(values)
        for value in val:
            val_ = []
            for _ in value:
                if _ in  value_drop:
                    continue
                val_.append(_)
            valu_.append(val_)
        result = []
        for i in valu_:
            result.append(' '.join(i))
        return result

# Khởi tạo mô hình và dự báo kết quả audio#
B1: Mã hoá audio file sang dạng numpy/tensor<br>
B2: Pretrain Model và dự báo kết quả,mô hình đã được train sẵn với hơn 180 phút training <br>
B3: Mapping IPA39 sang dạng IPA69 <br>
B4: Trả ra kết quả dự báo của file audio

In [ ]:
class FbPretrain():
    def __init__(self,path_):
        self.path = path_
        self.path_labs,self.path_wavs=ReadLabFile(path_).path_wavs_1()
        self.two_letter,self.IPA = self. corpus_index()
        
    def wav_to_numpy_array(self):
        list_array_from_wav = []
        for path in self.path_wavs:
            input_,sr  = librosa.load(self.path+'/'+path,sr = 16000)
            list_array_from_wav.append(input_)
        return list_array_from_wav
    
    def list_phonemes2list_IPA(self):
        list_predict_wav2list = []
        list_array_from_wav = self.wav_to_numpy_array()
        processor = Wav2Vec2Processor.from_pretrained("facebook/wav2vec2-xlsr-53-espeak-cv-ft")
        model = Wav2Vec2ForCTC.from_pretrained("facebook/wav2vec2-xlsr-53-espeak-cv-ft")
        i=0
        for array_from_wav in list_array_from_wav:
            i=i+1
            print(i)
            input_values = processor(array_from_wav, return_tensors="pt",padding=True).input_values
            logits = model(input_values).logits
            predicted_ids = torch.argmax(logits, dim=-1)
            predict_wav_list = processor.batch_decode(predicted_ids)
            predict_wav_list = predict_wav_list[0].replace('ː',':')
            predict_wav_list = predict_wav_list.replace('ɡ','g')
            predict_wav_list_ = list(predict_wav_list.split())
            list_predict_wav2list.append(predict_wav_list_)
        return list_predict_wav2list
    
    def corpus_index(self):
        data = pd.read_excel('/content/wav_file/Fb_phonemes.xlsx')
        two_letter = list(data['2-letter'])
        IPA=list(data['IPA'])
        return two_letter,IPA
    
    def predict_pretrain(self):
        list_phonemes = self.list_phonemes2list_IPA()
        two_letter,IPA = self.corpus_index()
        list_ =[]
        for phonemes in list_phonemes:
            idx =[]
            for i,phoneme in enumerate(phonemes):
                idx.append(IPA.index(phoneme))
            list_.append(idx)
        return list_
    
    def list_phonemes2list_L2corpus(self):
        lists_ = self.predict_pretrain()
        result_2gram=[]
        for list_ in lists_:
            value_2gram=[]
            for id_2gram in list_:
                value_2gram.append(self.two_letter[id_2gram])
            result_2gram.append(value_2gram)
        return result_2gram
             
    def list_phonemes2string_L2corpus(self):
        phonemes2str = []
        for result in self.list_phonemes2list_L2corpus():
            phonemes2str.append(' '.join(result))
        return phonemes2str

class Predict():
    def __init__(self,path_,model = 'Al'):
        self.path_ = path_
        self.path_labs,self.path_wavs = ReadLabFile(path_).path_wavs_1()
        self.model = model

    def predict_with_model(self):
        result =FbPretrain(self.path_).list_phonemes2string_L2corpus()
        return result

# Thẩm định độ chính xác thuật toán và hiển thị kết quả#
B1: Kiểm tra độ chính xác của thuật toán theo phương pháp per(phonemes erros),wer(word erros)<br>
B2: Hiển thị kết quả<br>

In [ ]:
class Score():
    def __init__(self,path_,model = 'Al'):
        self.path_ = path_
        self.hypothesis=Predict(path_,model).predict_with_model()
        self.ground_truth = ReadLabFile(path_).read_speech_file()
        
    def score(self):
        accuracy_wer_score =1- wer(self.ground_truth,self.hypothesis)
        accuracy_cer_score =1-cer(self.ground_truth,self.hypothesis)
        print('Accuracy with wer: ',accuracy_wer_score)
        print('Accuracy with cer: ',accuracy_cer_score)
        return accuracy_wer_score,accuracy_cer_score

    def display_result(self):
        hypothesis=self.hypothesis
        ground_truth=self.ground_truth
        name_lab,name_wav = ReadLabFile(self.path_).path_wavs_1()
        data = pd.DataFrame({'name_wav':name_wav,
                             'name_lab':name_lab,
            'hypothesis':hypothesis,
                      'ground_truth':ground_truth})
        return data

### Độ chính xác acuracy

In [ ]:
per ,wer = Score('/content/wav_file/abc').score()

In [ ]:
print('Accuracy with wer: ',wer)
print('Accuracy with cer: ',per)

Accuracy with wer:  0.9213759213759214
Accuracy with cer:  0.8725637181409296


## Hiển thị kết quả

In [ ]:
data = Score('/content/wav_file/abc').display_result()

In [ ]:
data.head(20)

name_wav          name_lab  \
0   arctic_a0001.wav  arctic_a0001.lab   
1   arctic_a0002.wav  arctic_a0002.lab   
2   arctic_a0003.wav  arctic_a0003.lab   
3   arctic_a0004.wav  arctic_a0004.lab   
4   arctic_a0005.wav  arctic_a0005.lab   
5   arctic_a0006.wav  arctic_a0006.lab   
6   arctic_a0007.wav  arctic_a0007.lab   
7   arctic_a0008.wav  arctic_a0008.lab   
8   arctic_a0009.wav  arctic_a0009.lab   
9   arctic_a0010.wav  arctic_a0010.lab   
10  arctic_a0011.wav  arctic_a0011.lab   
11  arctic_a0012.wav  arctic_a0012.lab   
12  arctic_a0013.wav  arctic_a0013.lab   
13  arctic_a0014.wav  arctic_a0014.lab   
14  arctic_a0015.wav  arctic_a0015.lab   
15  arctic_a0016.wav  arctic_a0016.lab   
16  arctic_a0017.wav  arctic_a0017.lab   
17  arctic_a0018.wav  arctic_a0018.lab   
18  arctic_a0019.wav  arctic_a0019.lab   
19  arctic_a0020.wav  arctic_a0020.lab   

                                           hypothesis  \
0   ow th er r ah v dh ax d ey n jh er t r ey l f ...   
1   n ao t ae t dh ih s p er t ih k y ih l er k ey...   
2   f ao dh ax t w eh n t iy ax th t ay m dh ae t ...   
3   l ao d b ah t ay m g l ae d t ax s iy y uw aa ...   
4                w ih l w iy eh v er f er g eh t ih t   
5   g ao d b l eh s ih m ay hh ao p ay l g ao ao n...   
6   ae n d y uw aa l w ey z w ao n t t ax s iy ih ...   
7   g ae d y ao r l eh t er k ey m jh ah s t ih n ...   
8   hh iy t er n d sh aa p l iy ae n d f ey s t g ...   
9   ay m p l ey ih ng aa s ih ng g əl hh ae n d ih...   
10  ih f ay eh v er n iy d ih d aa f ay t er ih n ...   
11  g r ae g s ax n sh ah v d b ae k hh ih z ch ey...   
12  hh iy w ah z aa hh eh d sh ao t er dh ae n hh ...   
13  n aw y ow r k ah m ih ng d aa n t ax b ih z n ...   
14      ih t s dh ax ao r ao r aa b ao r iy aa l ax s   
15  dh eh z f ao t ch er ch ax l aa r ay f uh l sh...   
16  f r ah m dh ae t m ao m ax n t hh ih z f r eh ...   
17                  dh eh r w ah z aa ch ey n jh n aw   
18  ay f aa l ow d dh ax l ay n ah v dh ax p r ax ...   
19  k l ah b z ae n d b aa l z ae n s ih t iy z g ...   

                                         ground_truth  
0   ao th er ah v dh ax d ey n jh er t r ey l f ih...  
1   n aa t ae t dh ih s p er t ih k y ax l er k ey...  
2   f ao r dh ax t w eh n t iy ax th t ay m dh ae ...  
3   l ao r d b ah t ay m g l ae d t ax s iy y uw a...  
4                w ih l w iy eh v er f er g eh t ih t  
5   g aa d b l eh s eh m ay hh ow p ay l g ow aa n...  
6   ae n d y uw ao l w ey z w aa n t t ax s iy ih ...  
7   g ae d y ao r l eh t er k ey m jh ah s t ih n ...  
8   hh iy t er n d sh aa r p l iy ae n d f ey s t ...  
9   ay m p l ey ih ng ax s ih ng g ax l hh ae n d ...  
10  ih f ay eh v er n iy d ax d ax f ay t er ih n ...  
11  g r eh g s ax n sh ah v d b ae k hh ih z ch eh...  
12  hh iy w aa z ax hh eh d sh ao r t er dh ae n h...  
13  n aw y ao r k ah m ih ng d aw n t ax b ih z n ...  
14        ih t s dh ax er ao r ax b ao r iy ae l ax s  
15  dh eh r z f ao r t ch er ch ih l ey r ay f ax ...  
16  f r ah m dh ae t m ow m ax n t hh ih z f r eh ...  
17                  dh eh r w aa z ax ch ey n jh n aw  
18  ay f aa l ow d dh ax l ay n ah v dh ax p r ax ...  
19  k l ah b z ae n d b ao l z ae n d s ih t iy z ...

#Kết Luận<br>
Ưu điểm:
Kết quả có độ chính xác khá cao, với *Accuracy* khoảng 92% dù chỉ là bản demo nhưng kết quả cũng rất tốt. Bản kiểm tra kết quả chính xác là tập dữ liệu của người Anh địa phương(Scotland)<br>
Nhược điểm: Thời gian tính toán mô hình chậm vào khoảng 1s 1 file. Thuật toán cần cải thiện thêm về độ chính xác và tốc độ tính toán.